v2/pkg/engine/resolve/resolve.go

```diff
@@ -689,8 +694,18 @@ func (r *Resolver) handleTriggerUpdate(id uint64, data []byte) {
 
 		select {
 		case <-r.ctx.Done():
+			// Skip sending all events if the resolver is shutting down
+			return
 		case <-c.ctx.Done():
+			// Skip sending the event if the client disconnected
 		case s.workChan <- fn:
+			// Send the work to the subscription worker
+		case <-s.completed:
+			// Stop sending if the subscription is completed. Otherwise, this could block the event loop forever
+			// when the subscription worker was shutdown after channel close but the event was still scheduled.
+			if s.resolver.options.Debug {
+				fmt.Printf("resolver:trigger:subscription:completed:%d:%d\n", s.id.ConnectionID, s.id.SubscriptionID)
+			}
```

Comment:
Is this for debugging? Shouldn't it be a structured log?

Reply:
That's how we did and doing it for now. I agree, it's not the best approach. There is a ticket in backlog to Zap for everything in the engine to align it with the router.

Reply:
if it's from scratch we should probably use https://pkg.go.dev/log/slog if it's an option

Reply:
Let's have this discussion outside of this PR.